src/uct/ib/ud/base/ud_ep.c

```diff
@@ -881,92 +1001,76 @@ static void uct_ud_ep_rx_ctl(uct_ud_iface_t *iface, uct_ud_ep_t *ep,
-                                             (uct_device_addr_t*)uct_ud_creq_ib_addr(creq));
-    if (status != UCS_OK) {
-        return NULL;
+    switch (ctl->type) {
+    case UCT_UD_PACKET_CREQ:
+        break;
+    case UCT_UD_PACKET_CREP:
+        uct_ud_ep_rx_crep(iface, ep, neth, ctl);
+        break;
+    case UCT_UD_PACKET_FIN:
+        /* Got FIN packet which means that the endpoint won't be used to
+         * to data anymore */
+        ucs_assert(!(ep->flags & UCT_UD_EP_FLAG_CONNECT_TO_EP));
+        uct_ud_ep_handle_peer_disconnect(ep);
+        break;
+    default:
+        ucs_fatal("unexpected control packet type - %d", ctl->type);
     }
+}
 
-    uct_ud_peer_name(ucs_unaligned_ptr(&creq->peer));
+static UCS_F_NOINLINE void
+uct_ud_ep_process_unexp_packet(uct_ud_iface_t *iface, uct_ud_ep_t *ep,
+                               uct_ud_neth_t *neth, unsigned length)
+{
+    char buf[256] = {0};
 
-    skb->len = sizeof(*neth) + sizeof(*creq) + iface->super.addr_size;
-    return skb;
+    uct_ud_dump_packet(&iface->super.super, UCT_AM_TRACE_TYPE_RECV, neth,
+                       length, length, buf, sizeof(buf) - 1);
+    ucs_fatal("ep=%p: unexpected non-AM packet: [%s]", ep, buf);
 }
 
 void uct_ud_ep_process_rx(uct_ud_iface_t *iface, uct_ud_neth_t *neth, unsigned byte_len,
                           uct_ud_recv_skb_t *skb, int is_async)
 {
+    uct_ud_ep_t *ep = NULL;
     uint32_t dest_id;
     uint32_t is_am, am_id;
-    uct_ud_ep_t *ep = 0; /* todo: check why gcc complaints about uninitialized var */
     ucs_frag_list_ooo_type_t ooo_type;
 
-    UCT_UD_IFACE_HOOK_CALL_RX(iface, neth, byte_len);
+    UCT_UD_IFACE_HOOK_CALL_RX(iface, neth, goto out);
 
     dest_id = uct_ud_neth_get_dest_id(neth);
     am_id   = uct_ud_neth_get_am_id(neth);
     is_am   = neth->packet_type & UCT_UD_PACKET_FLAG_AM;
 
     if (ucs_unlikely(dest_id == UCT_UD_EP_NULL_ID)) {
-        /* must be connection request packet */
-        uct_ud_ep_rx_creq(iface, neth);
-        goto out;
+        /* Must be connection request packet */
+        ep = uct_ud_ep_rx_creq(iface, neth);
+        if (ep == NULL) {
+            goto out;
+        }
+
+        goto frag_list_insert;
     } else if (ucs_unlikely(!ucs_ptr_array_lookup(&iface->eps, dest_id, ep) ||
-                            (ep->ep_id != dest_id)))
-    {
-        /* Drop the packet because it is
-         * allowed to do disconnect without flush/barrier. So it
-         * is possible to get packet for the ep that has been destroyed
+                            (ep->flags & UCT_UD_EP_FLAG_DISCONNECTED))) {
+        /* Drop the packet because it is allowed to do disconnect without
+         * flush/barrier. So it is possible to get packet for the ep that has
+         * already been destroyed
          */
-        ucs_trace("RX: failed to find ep %d, dropping packet", dest_id);
+        ucs_trace("RX: dropping packet dest_id=%u, ep=%p", dest_id, ep);
```

Comment:
IMO better print different error message for each case (ep not found, and ep disconnected)
then, can avoid initializing ep to NULL

Reply:
done